### ISSUE TITLE
Nvidia GPU Graphics Performance Architect Prep

### ISSUE THREAD

**Issue by Timothyoung97** [Timothyoung97/RenderingEngine#4]
##### 给定一个点，判断该点是否在三角形里面

To determine if a point is inside a triangle, you can follow these steps:
- Assume the vertices of the triangle are A, B, and C, and the point you want to check is P.
- Calculate the cross products of vectors AP, BP, and CP.
- If all the cross products have the same direction (i.e., their orientations are consistent), then the point P is inside the triangle. Otherwise, it's outside.

##### 光栅化时，如果多个三角形共享一个顶点，如何制定一个合理的规则保证每个顶点只被光栅化一次；

One common approach is to use a data structure like an edge list or an active edge table (AET) along with a scanline algorithm. Here's a simplified explanation of how this can work:

- Edge List: Maintain a list of edges for each triangle, where each edge is defined by its starting and ending vertices. Combine all edge lists into a single list, sorted by the edge's starting y-coordinate. This creates a sorted list of edges that the scanline algorithm can use.

- Active Edge Table (AET): As you process each scanline, update an active edge table with edges that intersect that scanline. Include information like the x-coordinate of the intersection point and the slope of the edge.

- Scanline Algorithm: For each scanline, traverse the active edge table and fill in the pixels between pairs of intersecting edges. Update the active edge table as edges enter or leave the scanline.

The edge list ensures that edges are processed in a consistent order, while the active edge table manages which edges are currently active for each scanline. 

##### 为何硬件绘制时通常都以三角形为单位而不是其它多边形；

- Simplicity and Efficiency: Triangles are the simplest polygon to represent and render. Rendering hardware is optimized for processing triangles efficiently. Many rendering algorithms, such as rasterization, interpolation, and shading, are naturally suited to triangles. Other polygons would require more complex algorithms and processing, which can decrease rendering performance.
- Planarity: Triangles are always planar, meaning they lie flat on a two-dimensional surface. This simplifies various computations such as surface normal calculations, lighting calculations, and texture mapping.
- Completeness: Triangles can tessellate any surface or shape. By combining multiple triangles, you can approximate any arbitrary shape with a high degree of accuracy. This property allows for versatile and flexible rendering of complex scenes.
- Interpolation: Triangles facilitate easy interpolation of attributes such as color, texture coordinates, and normals across their vertices. This interpolation is crucial for achieving smooth shading effects and realistic rendering.
- Mathematical Properties: Triangles have well-defined mathematical properties that simplify various geometric calculations, including intersection tests, collision detection, and visibility determination.

##### 知道三角形三个顶点的颜色，光栅化时如何计算三角形内部其它点的颜色；
- Barycentric Coordinates: Before interpolation, each pixel inside the triangle is mapped to the triangle using its barycentric coordinates. Barycentric coordinates express any point inside a triangle as a linear combination of its vertices. These coordinates are used to determine how much each vertex contributes to the color of the point.
- Interpolation: Once the barycentric coordinates of a pixel are known, the colors of the triangle's vertices are interpolated to find the color of the pixel. This interpolation can be linear, bilinear, or use more advanced techniques like perspective-correct interpolation depending on the requirements.
- Color Interpolation Formula: The color of each pixel is computed by interpolating the colors of the triangle's vertices weighted by the pixel's barycentric coordinates. The formula for interpolation typically involves linearly interpolating the red, green, blue, and possibly alpha components of the colors separately.
  <img width="521" alt="image" src="https://github.com/Timothyoung97/RenderingEngine/assets/62177572/c6adc10c-06d2-4a45-9f8a-e0b393d1236d">
- Clipping and Depth Testing: After computing the color for each pixel, additional operations such as clipping against the viewport and depth testing against the depth buffer may be performed to determine the final color that gets displayed on the screen.

###### Good
- Data Structures: For example, using a vector to store vertices or pixels in an image can provide efficient access and manipulation compared to other data structures.
- Algorithms: STL provides efficient algorithms for sorting, searching, and manipulating data. These algorithms can be beneficial in various graphics tasks such as sorting triangles or performing geometric calculations. For instance, algorithms like std::sort() can be used to sort objects based on certain criteria, which is essential in rendering for depth buffering and transparency sorting.
- Ease of Use: STL's ease of use and consistent interface can improve productivity and code maintainability, which indirectly contributes to efficiency. Developers can focus more on the logic of their algorithms rather than spending time implementing basic data structures and algorithms from scratch.
- Integration with Graphics APIs: STL can seamlessly integrate with graphics APIs like OpenGL or DirectX. Using STL containers to manage resources or data passed to and from the graphics pipeline can streamline development and potentially improve efficiency by leveraging the performance optimizations provided by these APIs.

###### Bad
- Performance Overhead: STL can introduce performance overhead due to its generic nature and dynamic memory allocation. In graphics and game programming, where real-time performance is critical, developers often seek more control over memory management and data structures to minimize overhead.
- Memory Fragmentation: STL's dynamic memory allocation and deallocation can lead to memory fragmentation over time. In scenarios where memory fragmentation can impact performance, such as long-running games or real-time graphics applications, developers may prefer more deterministic memory management strategies.
- Lack of Control: STL abstracts many implementation details, which can be beneficial for productivity but may also limit the level of control that developers have over performance-critical aspects of their code. Graphics and game programmers often need fine-grained control over memory layout, data access patterns, and algorithmic optimizations to achieve optimal performance.
- Platform Dependency: While STL is part of the C++ standard, different implementations (e.g., GCC's libstdc++, Microsoft's STL, etc.) may behave differently or have varying performance characteristics. In game development, where cross-platform compatibility is essential, developers may prefer to minimize reliance on platform-specific behavior or dependencies.
- Size and Complexity: Some components of STL, such as standard containers and algorithms, may have a relatively large code footprint and complexity. In constrained environments like embedded systems or mobile devices, where resource usage is tightly controlled, developers may opt for more lightweight or specialized libraries tailored to their specific needs.
- Debugging and Profiling Challenges: STL code can be more challenging to debug and profile compared to custom implementations or libraries with simpler internals. In performance-critical scenarios, such as optimizing rendering pipelines or physics simulations, having clear insights into code behavior and performance characteristics is crucial for identifying bottlenecks and making optimizations.

##### 写一个函数，分配大小是32字节倍数的内存；
```cpp
#include <cstdlib>
#include <iostream>

void* allocateMemory(size_t size) {
    // Calculate the nearest multiple of 32 bytes
    size_t alignedSize = ((size + 31) / 32) * 32;

    // Allocate memory using the standard library function malloc
    void* ptr = std::malloc(alignedSize);

    // Check if memory allocation was successful
    if (ptr == nullptr) {
        std::cerr << "Memory allocation failed." << std::endl;
        return nullptr;
    }

    return ptr;
}

int main() {
    // Example usage: allocate 50 bytes
    void* ptr = allocateMemory(50);

    // Check if allocation was successful
    if (ptr != nullptr) {
        std::cout << "Memory allocated successfully." << std::endl;
        // Don't forget to free the allocated memory when done using it
        std::free(ptr);
    }

    return 0;
}

```

##### 写一个屏幕拷贝的函数，将屏幕上的一片区域拷贝到令外一个地方；

- Can use compute shader, look up the targeted screen area, downsample or upsample to the result screen area

```cpp
#include <iostream>
#include <vector>

// Represents a rectangular region on the screen
struct ScreenRegion {
    int x, y;       // Top-left corner coordinates
    int width, height;  // Width and height of the region
};

// Function to copy a screen region to another location
void copyScreenRegion(const std::vector<std::vector<char>>& screen, const ScreenRegion& srcRegion, 
                      std::vector<std::vector<char>>& destination, int destX, int destY) {
    // Ensure source and destination regions are valid
    if (srcRegion.x < 0 || srcRegion.y < 0 || srcRegion.x + srcRegion.width > screen[0].size() ||
        srcRegion.y + srcRegion.height > screen.size() || destX < 0 || destY < 0 ||
        destX + srcRegion.width > destination[0].size() || destY + srcRegion.height > destination.size()) {
        std::cerr << "Invalid screen regions or destination coordinates." << std::endl;
        return;
    }

    // Copy the screen region to the destination
    for (int y = 0; y < srcRegion.height; ++y) {
        for (int x = 0; x < srcRegion.width; ++x) {
            destination[destY + y][destX + x] = screen[srcRegion.y + y][srcRegion.x + x];
        }
    }
}

int main() {
    // Example usage
    // Define screen dimensions
    const int screenWidth = 80;
    const int screenHeight = 24;

    // Create the screen buffer
    std::vector<std::vector<char>> screen(screenHeight, std::vector<char>(screenWidth, '.'));

    // Draw some characters on the screen (just for demonstration)
    for (int y = 5; y < 10; ++y) {
        for (int x = 10; x < 20; ++x) {
            screen[y][x] = '*';
        }
    }

    // Define the source screen region to copy
    ScreenRegion srcRegion = {10, 5, 10, 5};

    // Define the destination screen buffer
    std::vector<std::vector<char>> destination(srcRegion.height, std::vector<char>(srcRegion.width, '.'));

    // Copy the source screen region to the destination at coordinates (30, 10)
    copyScreenRegion(screen, srcRegion, destination, 30, 10);

    // Print the destination buffer to verify the copy operation
    for (int y = 0; y < srcRegion.height; ++y) {
        for (int x = 0; x < srcRegion.width; ++x) {
            std::cout << destination[y][x];
        }
        std::cout << std::endl;
    }

    return 0;
}

```

**Comment by Timothyoung97** [comment]
画一张GPU渲染流程图。
![image](https://github.com/Timothyoung97/RenderingEngine/assets/62177572/b33d50cb-778c-4739-a9b8-5c8b7ba235cd)

**Comment by Timothyoung97** [comment]
- Ask several questions about the graphcis pipeline and the overview of how the GPU execute, several questions about the graphics API usage, how the driver submit the commands to GPU. Then some very simple questions about data structure and algorithm.
- Describe how to use vulkan, and what should be take care of when create vulkan queues.

##### Overview of graphics pipeline
- Vertex Processing:
The process begins with vertex data, which defines the geometry of objects in 3D space.
Vertex shaders are executed for each vertex in the input geometry, transforming their positions from 3D world coordinates to 2D screen coordinates.
Other per-vertex attributes such as color, texture coordinates, and normals may also be processed in this stage.

- Primitive Assembly and Tessellation (Optional):
If tessellation is enabled, the GPU can dynamically subdivide geometric primitives (such as triangles) into smaller, more detailed primitives.
Tessellation control and evaluation shaders are executed to determine the level of tessellation and generate new vertices as needed.

- Geometry Processing:
In this stage, the GPU processes the geometry data to generate the final primitives that will be rasterized.
Geometry shaders (optional) can be used to perform additional transformations, generate new primitives, or apply procedural effects.

- Clipping and Culling:
Primitives that fall outside the view frustum are clipped, removing geometry that is not visible.
Back-face culling may also be applied to discard primitives that are facing away from the viewer.

- Rasterization:
The rasterization stage converts the primitives (usually triangles) into fragments/pixels that will be processed further.
For each triangle, the rasterizer generates fragments corresponding to pixels covered by the triangle on the screen.

- Fragment Processing:
Fragment shaders (also known as pixel shaders) are executed for each fragment generated by the rasterizer.
Fragment shaders compute the final color of each pixel, applying lighting, texturing, and other effects.
Depth testing and stencil testing may be performed to determine whether fragments are visible and should contribute to the final image.

- Per-sample Operations (Optional):
If multisampling is enabled, per-sample operations such as anti-aliasing and alpha blending may be applied to improve image quality.

- Framebuffer Operations:
Finally, the processed fragments are written to the framebuffer, which represents the image that will be displayed on the screen.
Additional operations such as depth buffering and alpha blending may be performed at this stage to combine fragments from multiple primitives.

##### API Questions
- Main differences between DirectX and Vulkan:
DirectX is a high-level graphics API developed by Microsoft primarily for Windows platforms, offering ease of use and extensive documentation. Vulkan, on the other hand, is a low-level, cross-platform API designed for high-performance graphics and compute applications.
Vulkan provides more fine-grained control over hardware resources and parallelism, making it suitable for highly optimized and efficient rendering. DirectX abstracts some of these details, making it easier to use but potentially less efficient in certain scenarios.

- Resource management in graphics API:
Graphics APIs provide functions for creating, binding, updating, and releasing resources such as vertex buffers, textures, and render targets.
Best practices for efficient resource utilization include minimizing redundant resource updates, batching draw calls to reduce state changes, and using resource pooling to reuse memory allocations.

- Setting up a rendering pipeline:
In DirectX 12 or Vulkan, setting up a rendering pipeline involves creating and configuring pipeline state objects, which specify the configuration of vertex input, vertex shaders, rasterization, pixel shaders, and output merge.
Key stages include pipeline creation, shader compilation, and binding pipeline state objects to the rendering context before issuing draw calls.

- Shader compilation and linking:
Graphics APIs typically provide shader compilation functions to compile shaders from source code into binary formats that can be executed by the GPU.
Debugging and optimizing shaders often involve using graphics debugger tools such as PIX for Windows or RenderDoc to analyze shader performance and identify rendering artifacts.

- Synchronization and resource barriers:
Graphics APIs provide mechanisms such as fences, semaphores, and synchronization primitives to control the execution order of rendering commands and prevent data hazards.
Resource barriers are used to synchronize access to resources between different stages of the graphics pipeline, ensuring correct ordering of memory accesses and preventing race conditions.

- Benefits and limitations of high-level vs. low-level APIs:
High-level APIs like DirectX and OpenGL provide abstraction layers that simplify development but may sacrifice some performance and flexibility compared to low-level APIs like Vulkan or Metal.
Low-level APIs offer more control over hardware resources and parallelism, enabling developers to optimize performance for specific hardware configurations and use cases.

- State management in graphics API:
Graphics APIs maintain state objects representing the configuration of rendering pipeline stages, resources, and rendering parameters.
Efficient state management involves minimizing state changes, batching draw calls with similar state configurations, and using state inheritance mechanisms to reuse common state configurations.

- Multi-threading and parallelism:
Graphics APIs support multi-threaded rendering by allowing multiple threads to issue rendering commands concurrently.
Strategies for optimizing rendering performance in a multi-threaded environment include minimizing synchronization overhead, parallelizing compute-intensive tasks, and distributing workloads across multiple CPU cores.

- Immediate mode vs. retained mode rendering:
Immediate mode rendering involves issuing rendering commands directly from the application's main loop, whereas retained mode rendering involves building and submitting rendering commands to a command buffer for later execution.
The choice between the two approaches depends on factors such as application complexity, performance requirements, and development preferences.

- Debugging rendering issues:
Debugging rendering issues involves using graphics debugger tools to analyze GPU events, shader execution, and rendering pipeline stages.
Techniques for diagnosing and fixing rendering artifacts or performance issues include inspecting shader code, capturing frame traces, and profiling GPU performance to identify bottlenecks.

##### From Driver to GPU
- API Abstraction:
Graphics applications interact with the GPU through high-level graphics APIs such as DirectX, OpenGL, or Vulkan.
These APIs provide a set of functions and data structures for creating rendering resources, specifying rendering commands, and managing the rendering pipeline.

- Command Generation:
The application's rendering logic generates commands in the form of API calls, specifying operations such as setting rendering states, binding resources (e.g., vertex buffers, textures), and issuing draw calls.
The commands are typically encapsulated into command lists or command buffers, which represent batches of rendering commands that can be executed by the GPU.

- Command Submission:
When the application is ready to submit rendering commands to the GPU, it calls the appropriate API function to submit the command list or command buffer.
The graphics driver intercepts these API calls and prepares the command data for submission to the GPU.

- Driver Translation:
The driver translates the high-level API commands into a format that is understandable by the GPU hardware.
This translation process involves mapping API calls to corresponding hardware-specific commands and data structures, as well as performing any necessary optimizations or transformations.

- Command Queueing:
The translated commands are then placed into command queues, which are managed by the graphics driver.
Command queues are used to organize and prioritize the execution of rendering commands, ensuring that they are processed in the correct order and according to the application's synchronization requirements.

- Command Buffer Submission:
Once the commands are queued, the driver initiates the submission process by sending the command buffers to the GPU.
This typically involves copying the command data from system memory to GPU memory, where it can be accessed and executed by the GPU.

- GPU Execution:
The GPU processes the submitted command buffers asynchronously, executing the specified rendering operations in parallel across multiple processing units.
The command processing pipeline within the GPU processes the commands sequentially, with each command being executed in the order they were submitted.

- Synchronization and Memory Access:
During command execution, the GPU may need to access various memory resources such as vertex buffers, textures, and render targets.
Synchronization mechanisms such as memory barriers and fences are used to ensure correct ordering of memory accesses and prevent data hazards.

- Completion Signaling:
Once all commands in a command buffer have been executed, the GPU signals completion back to the driver.
This allows the driver to perform any necessary cleanup tasks and prepare for the next frame or batch of rendering commands.

- Frame Presentation:
After rendering is complete, the final rendered frame is typically presented to the screen for display.
This involves swapping buffers, where the completed frame buffer is swapped with the front buffer to update the display with the latest rendered image.

##### DirectX11 
- Initialization:
Initialize the DirectX 11 device and device context.
Create a swap chain for presenting rendered frames to the screen.
Set up the viewport and other rendering settings.

- Resource Creation:
Create vertex buffers, index buffers, constant buffers, textures, and other resources needed for rendering.
Compile shaders (vertex, pixel, etc.) using HLSL and create shader objects from the compiled bytecode.
Set the input layout for vertex data.

- Rendering:
Clear the back buffer and depth/stencil buffer.
Set the vertex and pixel shaders, input layout, and other pipeline states.
Set shader constants and bind vertex and index buffers.
Issue draw calls to render geometry.
Present the rendered frame to the screen using the swap chain.

- Cleanup:
Release DirectX 11 resources and clean up memory allocations.
Release the swap chain and device objects.
Properly handle errors and exceptions to ensure graceful shutdown.

##### Understanding DX11 Command Queues
DirectX 11 doesn't have explicit command queues like newer APIs such as DirectX 12 or Vulkan. Instead, command submission is handled implicitly by the device context.

However, it's essential to understand that DirectX 11 devices can still benefit from multithreaded command submission by using deferred contexts.

- Considerations for Multithreaded Rendering:
If you're using deferred contexts for multithreaded rendering, ensure that resource access is synchronized properly to avoid data hazards.
Use synchronization mechanisms such as mutexes or semaphores to coordinate access to shared resources between multiple rendering threads.

- Resource Management:
Be mindful of resource lifetimes when using deferred contexts. Resources created in one deferred context must be used and released within the same context.
Avoid excessive resource creation and destruction in deferred contexts, as this can lead to increased overhead and reduced performance.

- Error Handling:
Implement robust error handling mechanisms to detect and handle errors that may occur during command submission or resource management.
Use DirectX debug layers and tools like PIX for Windows to diagnose and debug rendering issues.

- Performance Optimization:
Profile your application using performance analysis tools to identify bottlenecks and optimize rendering performance.
Experiment with different threading strategies and rendering techniques to achieve the best balance between CPU and GPU utilization.

**Comment by Timothyoung97** [comment]
Graphics performance team works on delivering an efficient and powerful graphics architecture every generation. The team studies graphics workloads and test out innovative HW/SW solutions on various platforms to address the inefficiencies in the current architecture. The work we do paves the path for real time rendering of some of the most complex and compute intensive visualization technique.

What you'll be doing: 
- Investigate and study state-of-the-art real-time rendering techniques their implementation on GPU, to find performance issues
- Study D3D12, DX Machine Learning, DX and Vulkan APIs and their impact to GPU
- Propose ideas to improve GPU architecture based on quantitative study of existing and projected architectures.
- Develop performance simulation models and simulation infrastructure.
- Design performance testplan and tests for new graphics units and architectural features.

What we need to see: 
- Bachlor in CS or EE. Master and PHD is preferred.
- Solid knowledge of computer science, with the focus on computer graphics and/or computer architecture.
- Strong interest in GPU graphics architecture, real time rendering technologies.
- Good mastery of C++ language and practical C/C++ development experience.

Ways to stand out from the crowd: 
- Good understanding of state-of-the-art rendering techniques and their usage of GPU.
- Experience of performance modelling, profiling and analysis is a plus.

##### Good understanding of state-of-the-art rendering techniques and their usage of GPU
- Ray Tracing: Ray tracing simulates the behavior of light in a scene by tracing the path of individual rays. Modern GPUs utilize hardware acceleration to perform ray tracing operations, enabling realistic lighting, reflections, and shadows in real-time rendering.
- Global Illumination: Techniques such as path tracing and photon mapping are used to accurately simulate indirect lighting effects, including diffuse inter-reflections and color bleeding. GPUs accelerate these computations through parallel processing.
- Screen Space Reflections (SSR): SSR is a technique used to simulate reflections by sampling information directly from the screen buffer. GPUs optimize the computation of reflections within the visible screen space, improving efficiency and performance.
- Deferred Rendering: Deferred rendering separates the rendering process into multiple passes, allowing for more efficient lighting calculations. GPUs efficiently handle the deferred rendering pipeline, enabling complex lighting effects with a large number of light sources.
- Tessellation: Tessellation divides polygons into smaller, more detailed primitives, enhancing the geometric complexity of objects in a scene. GPUs support hardware tessellation, allowing for the dynamic generation of detailed geometry on the fly.
- Compute Shaders: Compute shaders leverage the GPU's parallel processing capabilities to perform general-purpose computations. These shaders are used for tasks such as physics simulations, post-processing effects, and procedural generation, enhancing the realism and interactivity of rendered scenes.
- Subsurface Scattering: Subsurface scattering simulates the behavior of light as it penetrates and scatters within translucent materials like skin or wax. GPUs accelerate subsurface scattering calculations, enabling more realistic rendering of materials with complex light interaction.
- Volumetric Rendering: Volumetric rendering techniques simulate the interaction of light with volumetric data, such as clouds, smoke, or fog. GPUs optimize the rendering of volumetric effects through efficient sampling and ray marching algorithms.

##### 
- Performance Modeling:
Building mathematical or simulation models to predict the performance of a system under different workloads or configurations.
Analyzing factors such as CPU/GPU utilization, memory bandwidth, and latency to estimate system performance.
Using tools like queuing theory, regression analysis, or machine learning to develop predictive models.

- Performance Profiling:
Profiling software applications or algorithms to identify performance bottlenecks and hotspots.
Utilizing profiling tools such as Intel VTune, NVIDIA Nsight, or AMD CodeXL to collect data on CPU/GPU usage, memory access patterns, and execution time.
Analyzing profiling data to understand where resources are being spent and prioritize optimization efforts.

- Performance Analysis:
Analyzing performance metrics to understand system behavior under different conditions or configurations.
Conducting experiments to measure the impact of changes to software algorithms, hardware configurations, or system parameters on performance.
Identifying opportunities for optimization, such as reducing computational complexity, improving memory access patterns, or parallelizing workloads.

- Benchmarking:
Developing and executing benchmarks to measure the performance of software or hardware components.
Comparing performance metrics across different systems, architectures, or implementations.
Identifying outliers or anomalies in benchmark results and investigating potential causes.

- Optimization:
Implementing optimizations based on insights gained from performance modeling, profiling, and analysis.
Applying techniques such as algorithmic optimizations, parallelization, vectorization, or memory optimization to improve performance.
Iteratively testing and refining optimizations to achieve desired performance goals.

- Scalability Analysis:
Assessing the scalability of software systems or algorithms with increasing workload sizes or system resources.
Identifying scalability bottlenecks and proposing solutions to improve scalability.
Analyzing the trade-offs between scalability and performance in distributed or parallel computing environments.

**Comment by Timothyoung97** [comment]
多态和继承的关系，C++和C之间的关系区别，描述函数调用过程

**Comment by Timothyoung97** [comment]
C++的虚函数表是每个class一个还是每个实例一个
简单描述一下动态规化的算法

**Comment by Timothyoung97** [comment]
1-1: 两个Unit，中间有一个FIFO，要求你设计这个FIFO，每个时钟周期进行一次操
作，每次只能读入一个数据，每次也只能读出一个数据。Unit每个始终周期可以对FIFO
进行多次读写，Unit1进Unit2出。
1-2: 空间给三个点(x0, y0, z0),(x1, y1, z1),(x2, y2, z2)。一个平面方程，ax+by+
cz+d = 0，如何判断三点组成的三角形和次平面的空间关系。
2-1: 给出一个人的位置，设计一个系统，要求根据这个位置找到与他最接近的麦当劳。
包括算法如何，数据结构怎样，数据库怎么建之类的。
2-2: 一个球形苹果，一个虫子从一点钻进去，经过定长l钻出来，l<2r，r为苹果半径，
问是否可以将苹果切成两半，保证一半是好的，可以的话怎么切？

英文第一个问题问项目中遇到的困难或是犯过什么样的错误。第二个问的是如果老板对
你做的东西不满意，你怎么处理。

**Comment by Timothyoung97** [comment]
八道题目，第一题有六个小题，一道小题0.5分， 剩下的七道题目一题1分。全英文试卷，要求英文作答。 1. (1)what is an abstract class (2)myclass a;myclass b;b=a; myclass a;myclass b=a;区别 (3)stl有哪三类？ (4)容器有哪几种？ (5)virtual destruction作用？ (6).exe运行时缺少dll，问如何知道缺少什么样的dll？ 2.程序运行结果显示 3.程序改错题 class foo{ public: foo(int j){i=new int[j];} ~foo(){delete i;} private: int *i; } class derive:foo{ public: derive(char j){i=new int[j];} ~derive(){delete i;} private: char *i; } main() { foo *f=new foo(100); foo *d=new derive(200); *f=*d; delete f; delete d; } 4.描述虚拟存储系统及其作用 5_tlb（快表）的作用 6.从编程者的角度来讲，四路组相联和全相联哪个好？ 7.用户态和内核态的区别？ 8.经典的过桥问题 有四个人过桥。一个要1分钟，一个要2分钟，一个要5分钟，还有一个要10分钟。 桥一次只能过两个人。因为天黑了，过桥必须要手电筒，但是只有一杆。那么如 何让这4个人小于19分钟内过河？

**Comment by Timothyoung97** [comment]
##### 给出stack结构，利用stack完成queue的操作。
```
Class stack{
Void push(data);
Void pop(&data);
Bool isempty;}
```
写出：
```
Class queue{

}
```
![image](https://github.com/Timothyoung97/RenderingEngine/assets/62177572/660b4064-a7bd-49a7-b48c-53d6c2f23e58)
![image](https://github.com/Timothyoung97/RenderingEngine/assets/62177572/49128fcb-ca16-459c-b639-31d1c75a078e)

##### 一个链表，里面数字无序排列，要求给出代码，实现升序排序。
   Void sort(*head){

}

put into vector, write a sort with a lambda function that using the member for comparison

Ascending:
```
std::ranges::sort(mMyClassVector, [](const MyClass &a, const MyClass &b)
{ 
    return a.mProperty < b.mProperty; 
});
```
Descending:
```
std::ranges::sort(mMyClassVector, [](const MyClass &a, const MyClass &b)
{ 
    return a.mProperty > b.mProperty; 
});
```

##### a=b*c+d; b,c,d均为unsigned 8bit，问a需要多大bit来存储。给出思考过程。

For unsigned 8-bit integers:
The range of values is from 0 to 255 (2^8 - 1).

- Step 1: Find the range of possible values for each operand:
b, c, d: Each can take values from 0 to 255.

- Step 2: Determine the maximum possible value for the expression b * c + d:
The maximum value of b * c occurs when both b and c are 255 (the maximum value for an 8-bit unsigned integer), which is 255 * 255 = 65025.
Adding the maximum value of d (255) to this product gives a maximum value of 65280.

- Step 3: Calculate the number of bits needed to represent the maximum value:
The maximum value 65280 can be represented using 16 bits (2^16 - 1 = 65535), which is greater than the maximum value that can be represented by 8 bits (255).
Therefore, we need at least 16 bits to store the result a.

- Conclusion:
The result a needs at least 16 bits to store the values produced by the expression b * c + d when b, c, and d are unsigned 8-bit integers.

##### 解释mipmap
You know this

##### 数据发送端：100clock中工作 80clock,休息 20clock, 但这80是random分布的。（1bit/1clock. )
数据接收端：每10clock中前 8clock 工作，后2clock休息。

问：作为中间的一个缓冲器，其容量应该为多大？

- Sender's Data Production Rate:
The sender works for 80 out of every 100 clock cycles, producing data continuously during this time.
The rate of data production is therefore 80 bits per 100 clock cycles, or 0.8 bits per clock cycle on average.
- Receiver's Data Consumption Rate:
The receiver works for 8 out of every 10 clock cycles, consuming data continuously during this time.
The rate of data consumption is therefore 8 bits per 10 clock cycles, or 0.8 bits per clock cycle on average.
- Buffering Requirements:
Since the sender and receiver have the same average data rate, the buffer size only needs to accommodate fluctuations due to the random distribution of data production.
The sender produces data at a rate of 0.8 bits per clock cycle on average, but this rate may vary due to the random distribution.
The buffer should be able to handle the maximum possible variation in data production within a given time frame.
- Calculation:
During the 80 clock cycles of work, the sender produces 80 * 0.8 = 64 bits on average.
However, due to the random distribution, the actual amount of data produced during these 80 clock cycles may vary.
To accommodate the maximum possible variation, the buffer size should be equal to or greater than the maximum amount of data that can be produced during 80 clock cycles.
Assuming the maximum variation occurs when the sender produces data continuously for 80 clock cycles, the buffer size should be at least 64 bits.
- Conclusion:
The buffer capacity should be at least 64 bits to accommodate the maximum possible variation in data production by the sender. This ensures that no data is lost due to temporary fluctuations in the sender's output.

##### 一个格子图，大概如下：

b   w   b   w   b
w   b   w   b   w
b   w   b   w   b
w   b   w   b   w
b   w   b   w   b
(1) 有多少个正方形？
(2) 有多少个方形（包括长方形，正方形）？
(3) 给你一个点，你如何判断它是黑色还是白色？写c代码。以左下角为原点。
注：b表示黑色，w表示白色。（上面所有小方格都是正方形：）。

**Comment by Timothyoung97** [comment]
第一面是software组的一个gg，还不错，可能是浙大的，因为他认识我老板。主要是根据
我简历写的skills来问的，涉及的范围比较广，C++，操作系统，体系结构，还有部分智
力题，历时一个多小时，期间2道题没给出完整答案。个人很喜欢这个gg的问题，感觉他
和我的兴趣差不多，所以题目我都比较熟悉，这个gg还不错，面试结束以后听说我中饭
没吃到处给我找东西，可惜只找到可乐，于是继续饿肚子。
接着是第二面，graphics组的面试官一开始就让我谈了谈做的项目，然后给了道坐标变
换的题目，很久没接触有点生疏，最后答出来，不过不是最好的，还没进入状态，算是
被鄙视了，后来的表现自认为很不错，主要问算法，还有GPU architecture相关的问题
，印象中都答出来了，算法问的也不深，比如找BST的两个也节点的最近公共父节点，
给了个算法，他比较满意，让我把程序写出来看看我的动手能力，也算过关了，还有问
到写数独的问题，总体上觉得很简单。历时一个多小时。
最后一个是graphics组的经理，主要涉及GPU，还有我的一些个人爱好。谈到个人爱好的
时候我跟他说我有很多兴趣方向，还谈了谈对以后的搜索技术趋势的理解。他让我最好
确定一个发展方向。历时1个小时的样子。

**Comment by Timothyoung97** [comment]
请找到这样的一个8位二进制数：当把它解释为一个十进制数或者二进制补码数的时候，它都具有相同的负值。

**Comment by Timothyoung97** [comment]
Several long "skype" interviews about computer architecture, physical-based rendering, graphics API, "life of triangle", C++, algorithms, path-tracing , screen space shaders, (bvh, importance sampling) , linear algebra and analytical geometry,
Pipeline based architecture of CPU
GPU architecture components
Life of triangle

**Comment by Timothyoung97** [comment]
rendering pipeline, minimapping, tessellation

**Comment by Timothyoung97** [comment]
how to find if a number of a power of 2 in constant time?

```
// Function to check if x is power of 2
bool isPowerOfTwo(int n)
{
    if (n == 0)
        return false;
 
    return (ceil(log2(n)) == floor(log2(n)));
}
```

**Comment by Timothyoung97** [comment]
The phone interview lasted 1 hours and was mostly OS and computer architectures problems.

directives, difference bewteen global and static in C, difference between a writeback and a writethrough cache, difference between kernel and user mode, malloc, ...

It was very hard-core. The interviewer thanked me for not bs-ing him and just telling him I don't know something.

**Comment by Timothyoung97** [comment]
Given two numbers a and b, we have an averaging function (a+b)/2. This function wont work under certain cases. I was supposed to tell the case, and also how to handle it.

given an unsigned char c, swap every two bits of it

**Comment by Timothyoung97** [comment]
##### 什么时候用Virtual Destructor
- Polymorphic Behavior:
If a class is intended to be a base class with polymorphic behavior (i.e., it has at least one virtual function), it should typically have a virtual destructor.
This ensures that when an object of a derived class is destroyed through a pointer to the base class, the appropriate destructor is called based on the dynamic type of the object.

- Memory Leaks and Undefined Behavior:
Without a virtual destructor, deleting an object of a derived class through a pointer to the base class may result in undefined behavior.
This can lead to memory leaks if the destructor of the derived class is not called, potentially leaving resources allocated by the derived class in an unreleased state.

- Proper Resource Cleanup:
A virtual destructor allows derived classes to properly clean up any resources they own before being destroyed.
For example, if a derived class allocates memory or opens a file, its destructor can release that memory or close the file, ensuring proper resource management.
Here's an example illustrating when to use a virtual destructor:

```cpp
class Base {
public:
    virtual ~Base() { } // Virtual destructor

    // Other virtual functions and non-virtual functions
};

class Derived : public Base {
public:
    ~Derived() {
        // Cleanup resources owned by Derived
    }

    // Other member functions
};
```
In this example, Base has a virtual destructor because it serves as a base class with polymorphic behavior. Derived inherits from Base and overrides the destructor to provide proper resource cleanup specific to Derived objects. When a Derived object is destroyed through a pointer to Base, the virtual destructor in Base ensures that the destructor of Derived is called.

##### Compiler对Virtual怎么处理
- Virtual Function Table (vtable):
The compiler creates a virtual function table (vtable) for each class that declares one or more virtual functions.
The vtable is an array of function pointers, where each entry corresponds to a virtual function declared in the class.
Each object of a class with virtual functions contains a hidden pointer to its corresponding vtable.

- Virtual Function Pointer (vptr):
Along with the vtable, the compiler adds a hidden virtual function pointer (vptr) to each object of a class with virtual functions.
The vptr points to the beginning of the object's vtable.

- Dynamic Dispatch:
When a virtual function is called through a pointer to the base class, the compiler uses the object's vptr to determine the correct function to call at runtime.
This process is known as dynamic dispatch or late binding because the decision about which function to call is made at runtime based on the actual type of the object.

- Overhead:
Adding virtual functions and vtables introduces some overhead in terms of memory consumption and runtime performance.
Each object with virtual functions requires additional memory to store the vptr, and there may be a slight performance penalty when invoking virtual functions due to the extra level of indirection required.

- Optimizations:
Compilers may apply various optimizations to reduce the overhead of virtual function calls.
For example, in some cases where the compiler can determine the exact type of the object at compile time (e.g., when calling a virtual function on a local object), it may be able to bypass the vtable lookup and directly call the appropriate function.

##### 最后一道coding，用google doc做的，给16-bit RGB值，转成32-bit RGBX值
```
// Function to convert 16-bit RGB value to 32-bit RGBX value
uint32_t convertRGB16to32(uint16_t rgb16) {
    // Extract 5-bit components from 16-bit RGB value
    uint8_t r5 = (rgb16 >> 11) & 0x1F;  // 5 bits for red
    uint8_t g6 = (rgb16 >> 5) & 0x3F;   // 6 bits for green
    uint8_t b5 = rgb16 & 0x1F;          // 5 bits for blue

    // Expand 5-bit components to 8 bits
    uint8_t r8 = (r5 * 255) / 31;  // Scale to 8 bits
    uint8_t g8 = (g6 * 255) / 63;  // Scale to 8 bits
    uint8_t b8 = (b5 * 255) / 31;  // Scale to 8 bits

    // Create 32-bit RGBX value (0xFF for alpha channel)
    uint32_t rgbx32 = (uint32_t)r8 << 24 | (uint32_t)g8 << 16 | (uint32_t)b8 << 8 | 0xFF;

    return rgbx32;
}
```
解释点乘、叉乘用途
Z-buffer
Z-fighting
Z-buffer用什么数据类型，大小，取舍
Depth testing
Stencil Buffer
Deferred shading和forward shading
听没听说过TBDR (Tile-Based Deferred Rendering)
讲讲Ray Tracing

**Comment by Timothyoung97** [comment]
怎么用函数拷贝很大的vector（unique_ptr+trailing return type)。

三个smart pointer，unique_ptr用等号赋值会发生什么，share_ptr和weak_ptr的区别。
设计一个算不同形状面积的class （虚函数，虚构造/析构函数执行顺序，friend class和derived的区别）
存在一个无限长的数组，开头到某个地方是有序的，其他全是-1 (e.g. [1,3,5,7,9,-1,-1,-1,-1,...])，如何查找特定数是否存在（先double stride找-1, 然后迂回找end, 然后bisect, 总复杂度log(n))
在不给head，只给当前linkedlist指针的情况下如何删掉这个指针（我没想到，面试官提示可以把后一个指针代替当前的指针赋值，我没答上来，估计跪在了这里。）

**Comment by Timothyoung97** [comment]
<img width="567" alt="image" src="https://github.com/Timothyoung97/RenderingEngine/assets/62177572/28100b72-047b-428c-806b-ac9551f13c05">

**Comment by Timothyoung97** [comment]
![image](https://github.com/Timothyoung97/RenderingEngine/assets/62177572/be407267-d126-44e2-974c-d6f911b3058d)

**Comment by Timothyoung97** [comment]
![image](https://github.com/Timothyoung97/RenderingEngine/assets/62177572/d2494fda-2f8e-44c6-8f76-548882054ec6)

**Comment by Timothyoung97** [comment]
![image](https://github.com/Timothyoung97/RenderingEngine/assets/62177572/c8e95d91-0a7d-4dd6-90d1-cf5790d611e9)

**Comment by Timothyoung97** [comment]
![image](https://github.com/Timothyoung97/RenderingEngine/assets/62177572/d137e29c-55a1-4600-bd6a-0fb22c37e9b2)

**Comment by Timothyoung97** [comment]
![image](https://github.com/Timothyoung97/RenderingEngine/assets/62177572/a2252810-99fd-40f9-8c3b-61e31b055ee3)

**Comment by Timothyoung97** [comment]
Can you elaborate on some of the specific challenges the graphics performance team has encountered in previous projects, and how these challenges were addressed?

How does the team prioritize between optimizing existing architecture and introducing new features in each generation of GPU architecture?

Could you provide examples of real-world applications or industries where the advancements in GPU architecture directly impact performance or efficiency?

How does the team ensure compatibility and performance across different APIs such as D3D12, DX Machine Learning, DX, and Vulkan?

What methodologies or tools does the team employ to quantify and analyze the performance of existing and projected architectures?

Can you discuss any recent innovations or breakthroughs in real-time rendering techniques that the team has been investigating or implementing?

How does the team balance between theoretical performance gains and practical implementation feasibility when proposing ideas to improve GPU architecture?

What role does collaboration with other teams, such as software development or hardware engineering, play in the process of improving GPU architecture?

Could you walk me through the typical process of developing performance simulation models and infrastructure within the graphics performance team?

Can you provide insights into the approach the team takes in designing performance test plans and tests for new graphics units and architectural features?
